NebulaPatcher/Patches/Dynamic/PlanetFactory_Patch.cs

```diff
@@ -716,4 +718,93 @@ public static void EntityFastFillIn_Postfix(PlanetFactory __instance, int entity
             stationComponent.idleShipCount = __state.Item4;
         }
     }
+
+    #region Combat
+
+    [HarmonyPrefix]
+    [HarmonyPatch(nameof(PlanetFactory.KillEnemyFinally))]
+    public static bool KillEnemyFinally_Prefix(PlanetFactory __instance, int enemyId)
+    {
+        if (!Multiplayer.IsActive || enemyId <= 0)
+        {
+            return true;
+        }
+        if (Multiplayer.Session.IsServer)
+        {
+            var starId = __instance.planet.star.id;
+            Multiplayer.Session.Network.SendPacketToStar(new DFGKillEnemyPacket(__instance.planetId, enemyId), starId);
+            return true;
+        }
+        if (Multiplayer.Session.Combat.IsIncomingRequest.Value)
+        {
+            return true;
+        }
+
+        // Client: wait for server to approve the unitId and enmeyId recycle
+        // Make this enemyData appear as empty
+        ref var enemyPtr = ref __instance.enemyPool[enemyId];
+        enemyPtr.isInvincible = true;
+        enemyPtr.id = 0;
+        Multiplayer.Session.Network.SendPacket(new DFGKillEnemyPacket(__instance.planetId, enemyId));
+
+        return false;
+    }
+
+    [HarmonyPrefix]
+    [HarmonyPatch(nameof(PlanetFactory.RemoveEnemyFinal))]
+    public static bool RemoveEnemyFinal_Prefix()
+    {
+        if (!Multiplayer.IsActive || Multiplayer.Session.IsServer)
+        {
+            return true;
+        }
+        // Only execute RemoveEnemyFinal when server approve on client
+        // Factories.IsIncomingRequest is for RemoveBasePit called on putting down Geothermal Power Station
+        return Multiplayer.Session.Combat.IsIncomingRequest.Value || Multiplayer.Session.Factories.IsIncomingRequest.Value;
+    }
+
+    [HarmonyPrefix]
+    [HarmonyPatch(nameof(PlanetFactory.KillEntityFinally))]
+    public static bool KillEntityFinally_Prefix(PlanetFactory __instance, int objId)
+    {
+        if (!Multiplayer.IsActive || objId <= 0)
+        {
+            return true;
+        }
+
+        if (Multiplayer.Session.IsClient) // Let server decide when to kill entity
+        {
+            return Multiplayer.Session.Factories.IsIncomingRequest.Value;
+        }
+
+        var packet = new KillEntityRequest(__instance.planetId, objId);
+        var starId = __instance.planet.star.id;
+        Multiplayer.Session.Server.SendPacketToStar(packet, starId);
+        return true;
+    }
+
+    [HarmonyPrefix]
+    [HarmonyPatch(nameof(PlanetFactory.ReconstructTargetFinally))]
+    public static void ReconstructTargetFinally_Prefix(PlanetFactory __instance, int prebuildId)
+    {
+        if (!Multiplayer.IsActive || Multiplayer.Session.Factories.IsIncomingRequest.Value)
+        {
+            return;
+        }
```

Comment:
Minor nitpick, could be simplified to:
```suggestion
        if (!Multiplayer.IsActive || Multiplayer.Session.Factories.IsIncomingRequest.Value) return;
```